### ISSUE TITLE
Schrödinger's resource is both a file and a directory

### ISSUE THREAD

**Issue by jimwise** [chef/chef#9175]
## Description

The `directory` resource only checks path existence and permissions, not that the resource target actually is a directory.  Thus, the replication recipe below, which creates (1) a file and then (2) a directory at the same path succeeds and leaves a file behind, when it should (non-idempotently) end up with a directory at that path.

## Chef Version

```
lindon$ chef-client --version
Chef Infra Client: 15.6.10
lindon$ 
```

## Platform Version

```
lindon$ cat /etc/centos-release
CentOS Linux release 7.7.1908 (Core)
lindon$ 
```

## Replication Case

The following self-contained recipe demonstrates the bug:
```
file '/var/tmp/test-chef-bug' do
  mode '0755'
  owner 'root'
  group node['root_group']
  content 'test'
  action :create
end

directory '/var/tmp/test-chef-bug' do
  mode '0755'
  owner 'root'
  group node['root_group']
  action :create
end
```

## Client Output

```
lindon$ sudo chef-client
Starting Chef Infra Client, version 15.6.10
[...]
Chef Infra Client finished, 8/475 resources updated in 13 seconds
lindon$ ls -l /var/tmp/test-chef-bug 
-rwxr-xr-x. 1 root root 4 Dec 17 20:02 /var/tmp/test-chef-bug
lindon$ 
```

## Stacktrace

No stack trace -- chef-client run completes without error.

**Comment by skippyj** [comment]
If I'm reading this right, I think it was a pretty easy fix? My attached PR just adjusts the create/delete action blocks from evaluating `unless File.exist?` to `unless File.directory?`

**Comment by jimwise** [comment]
That was my read, too — just making sure I’m not missing some reason this was done this way. 

Jim Wise
jwise@draga.com

Sent from my iPhone.
Digital signature available upon request.

**Comment by snehaldwivedi** [comment]
@lamont-granquist just wanted to confirm do you want here to remove files if created with same name and then create a directory & vice versa? OR it should create both if the recipe is as given in the description?

**Comment by jimwise** [comment]
I think it should be an error if a file with the same name exists — it’s easy enough to remove a file if that’s intentional, but if we remove a file, that could lose something which was meant to be mv’ed into the directory we are now creating.
-- 
				Jim Wise
				***@***.***

**Comment by snehaldwivedi** [comment]
@jimwise It sounds fine to me also and I was also working in the same way but the name of the issue `Directory resource should remove files and replace with a directory` describes something else so just waiting for confirmation to move forward.

**Comment by jimwise** [comment]
Yup.  It’s been a while — have given this more thought since then. :-)
-- 
				Jim Wise
				***@***.***

**Comment by snehaldwivedi** [comment]
@lamont-granquist can you please confirm which way we should move forward. Like as @jimwise mention removing files can lose something important. So we can add a msg like a file or a directory already exists or can create both (not sure if it's possible will need to check)?

**Comment by marcparadise** [comment]
Still waiting for @lamont-granquist 's thoughts, but I would expect in this case that it would fail in the second resource because the path exists and it is not of the expected type (dir).  Otherwise, I agree we risk overwriting /losing data. 

For a parallel example, if our recipe looks like this: 

```
directory "/tmp/test-link"

link "/tmp/test-link" do
  to "/usr"
end

```

The result is: 

```
    Errno::EISDIR
    -------------
    Is a directory @ apply2files - /tmp/test-link

```

**Comment by snehaldwivedi** [comment]
@marcparadise So if I have understood correctly, then I shall move forward to remove the file and create a directory?
